File: static/js/voyager.js
/*
	voyager.js

	This is a simple service which provides a way to archive and view adventure/travel ideas.
	@license Copyright (C) 2018  Joshua Reed
	This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

//sp_content and sp_local are both path specifiers from the server.
function init(sp_local, sp_content)
{
	console.log("Global init (untether_edit.js).");
	//initalize page. This is the global reference which we will call upon throughout the code.
	console.log("Using content root: " + sp_content);
	console.log("Using local root: " + sp_local);
	SP_LOCAL = sp_local;
	SP_CONTENT = sp_content;
	
	//If any ajax request returns an error, this is triggered.
	$(document).ajaxError(function(event, jqxhr, settings, thrownError)
	{
		console.log("===========XHR ERROR===========")
		console.log(jqxhr.status + ": " + jqxhr.responseText)
		console.log("===============================")
	});
	$.ajaxSetup({
		//Disable caching for projects. If a project is updated it will not show up even if the file name
		//changes due to caching. It's a pain, and we disable it here. This isn't exactly enterprise level
		//work, so who cares about performance.
		//cache: false
	});
	PAGE = new Page()
}

class Page
{
	//A general note about the function of this page.
	
	constructor()
	{
		$(window).resize(function ()
		{
			//Note, this can be called continuously or at the end of a resize, depending on browser.
			PAGE.onResize();
		});
		this.onResize();
		
		var _this = this;
		this.countries = []; //A list of all countries: [{country: name, states: [name, name, name]}]
		this.$loc_country = $('#loc_country');
		this.$loc_region = $('#loc_region');
		
		//Initialize the country selection dropdowns
		$.getJSON(SP_LOCAL + "/js/countrystate.json", function(json)
		{
			_this.countries = json.countries;
			for(var x = 0; x < json.countries.length; x++)
			{
				var c = json.countries[x];
				var $opt = $("<option value=" + x + ">" + c.country + "</option>");
				_this.$loc_country.append($opt);
			}
		});
		this.$loc_country.change(function()
		{
			_this.$loc_region.find('.vy_gen').each(function()
			{
				$(this).remove();
			});
			if(_this.$loc_country.val() == 'SELECT')
			{
				_this.$loc_region.hide();
				return;
			}
			var c = _this.countries[_this.$loc_country.val()];
			_this.populate_records(c.country, 'VY_ALL');
			if(c.states.length > 0)
			{
				for(var x = 0; x < c.states.length; x++)
				{
					var $opt = $("<option class='vy_gen' value=" + x + ">" + c.states[x] + "</option>");
					_this.$loc_region.append($opt);
				}
				_this.$loc_region.show();
			}
			else
			{
				_this.$loc_region.hide();
			}
		});
		this.$loc_region.change(function()
		{
			var code = "VY_ALL";
			var c = _this.countries[_this.$loc_country.val()];
			if(_this.$loc_region.val() != "VY_ALL")
			{
				code = c.states[_this.$loc_region.val()];
			}
			_this.populate_records(c.country, code);
		});
	}
	
	//Called when the window resizes
	onResize()
	{
		
	}
	
	//Populates records for a specific region or country. Country name should be passed, as well as region name or VY_ALL for all.
	populate_records(country, region)
	{
		console.log(country + " " + region);
		$.getJSON("/voyager/query/records", {country: country, region: region}, function(json)
		{
			console.log(json);
		});
	}
}

File: voyager.py
#voyager.py
#
#This is a simple service which provides a lightweight way to record and view interesting places and adventures across the world.
#Copyright (C) 2017  Joshua Reed
#This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.

from the_root.extensions import db, csrf
from the_root import app
from the_root.models import VoyagerType, VoyagerRecord
from the_root.decorators import render_template_standard

import flask
from flask_login import login_required, current_user

from flask import render_template, redirect, request, jsonify

@app.route("/voyager", methods=['GET', 'POST'])
@login_required # This must ALWAYS go below the route decorator! Otherwise anyn users can log in
def voyager():
	return render_template_standard("/voyager/overview.html")
	
@app.route("/voyager/add", methods=['GET', 'POST'])
@login_required # This must ALWAYS go below the route decorator! Otherwise anyn users can log in
def voyager_record():
	return render_template_standard("/voyager/record.html")
	
	
@csrf.exempt # This is added to allow post commands to access this method.
@app.route("/voyager/query/<query>", methods=['GET', 'POST'])
@login_required # This must ALWAYS go below the route decorator! Otherwise anyn users can log in
def voyager_query(query):
	
	if(query == 'records'):
		country = request.values.get('country', type=str)
		region = request.values.get('region', type=str)
		if(country == None or (region == '')):
			return "No name provided", 404
	elif(query == 'types'):
		types = VoyagerType.query.filter_by(user=current_user.id)
		t_list = []
		for type in types:
			t_list.append(type.getDict())
		return jsonify(t_list), 200
	elif(query == 'new_record'):
		name = request.values.get('name', type=str)
		location_type = request.values.get('location_type', type=str)
		new_type = request.values.get('new_type', type=int)
		country = None
		state = None
		lat = None
		lon = None
		if(location_type == 'reg'):
			country = request.values.get('country', type=str)
			state = request.values.get('state', type=str)
		elif(location_type == 'map'):
			pass
		else:
			lat = request.values.get('lat', type=float)
			lon = request.values.get('lon', type=float)
		if(new_type):
			type = request.values.get('type', type=str) # Name of new type
		else:
			type = request.values.get('type', type=int) # ID of existing type
		if(name == None or location_type == None or new_type == None or (country == None and lat == None and lon == None) or type == None):
			return "Some parameter was not provided", 404
		if(new_type):
			# Make the new type
			ntype = VoyagerType(type, current_user.id)
			db.session.add(ntype)
			db.session.commit()
			type = ntype.id # Now that we have an id, set the type so that the latter section of this code can make a record w/ type id.
		nrecord = VoyagerRecord(name, current_user.id, type, country, region=state, latitude=lat, longitude=lon)
		db.session.add(nrecord)
		db.session.commit()
		return jsonify({}), 200
		
	return '', 404